mp4/samplegroupentries.go

```diff
@@ -149,3 +152,147 @@ func (s *UnknownSampleGroupEntry) Info(w io.Writer, specificBoxLevels, indent, i
+	entry := &RapSampleGroupEntry{}
+	byt := sr.ReadUint8()
+	entry.NumLeadingSamplesKnown = byt >> 7
+	entry.NumLeadingSamples = byt & 0x7F
+	return entry, nil
+}
+
+func (s *RapSampleGroupEntry) Type() string {
+	return "rap "
+}
+
+func (s *RapSampleGroupEntry) Size() uint64 {
+	return 1
+}
+
+func (s *RapSampleGroupEntry) Encode(sw *SliceWriter) {
+	var byt uint8
+	byt |= (s.NumLeadingSamplesKnown << 7)
+	byt |= (s.NumLeadingSamples)
+	sw.WriteUint8(byt)
+}
+
+// Info - write box info to w
+func (s *RapSampleGroupEntry) Info(w io.Writer, specificBoxLevels, indent, indentStep string) (err error) {
+	bd := newInfoDumper(w, indent, s, -2)
+	bd.write(" * numLeadingSamplesKnown: %d", s.NumLeadingSamplesKnown)
+	bd.write(" * numLeadingSamples: %d", s.NumLeadingSamples)
+	return bd.err
+}
+
+// AlstSampleGroupEntry - Alternative Startup Entry - Grouping Type "alst"
+type AlstSampleGroupEntry struct {
+	RollCount         uint16
+	FirstOutputSample uint16
+	SampleOffset      []uint32
+	NumOutputSamples  []uint16
+	NumTotalSamples   []uint16
+}
+
+func (s *AlstSampleGroupEntry) Type() string {
+	return "alst "
+}
+
+func (s *AlstSampleGroupEntry) Size() uint64 {
+	// RollCount: 2
+	// FirstOutputSample: 2
+	// SampleOffset: 4 * count
+	// NumOutputSamples: 2 * count
+	// NumTotalSamples: 2 * count
+	return uint64(4 + 4*len(s.SampleOffset) + 2*len(s.NumOutputSamples) + 2*len(s.NumTotalSamples))
+}
+
+func DecodeAlstSampleGroupEntry(name string, length uint32, sr *SliceReader) (SampleGroupEntry, error) {
+	entry := &AlstSampleGroupEntry{}
+	entry.RollCount = sr.ReadUint16()
+	entry.FirstOutputSample = sr.ReadUint16()
+	entry.SampleOffset = make([]uint32, int(entry.RollCount))
+	for i := 0; i < int(entry.RollCount); i++ {
+		entry.SampleOffset[i] = sr.ReadUint32()
+	}
+
+	remaining := int(length-uint32(entry.Size())) / 4
+	if remaining <= 0 {
+		return entry, nil
+	}
+
+	// Optional
+	entry.NumOutputSamples = make([]uint16, remaining)
+	entry.NumTotalSamples = make([]uint16, remaining)
+	for i := 0; i < remaining; i++ {
+		entry.NumOutputSamples[i] = sr.ReadUint16()
+		entry.NumTotalSamples[i] = sr.ReadUint16()
+	}
+
+	return entry, nil
+}
+
+func (s *AlstSampleGroupEntry) Encode(sw *SliceWriter) {
+	sw.WriteUint16(s.RollCount)
+	sw.WriteUint16(s.FirstOutputSample)
+	for _, offset := range s.SampleOffset {
+		sw.WriteUint32(offset)
+	}
+	for i := range s.NumOutputSamples {
+		sw.WriteUint16(s.NumOutputSamples[i])
+		sw.WriteUint16(s.NumTotalSamples[i])
+	}
+}
+
+// Info - write box info to w
+func (s *AlstSampleGroupEntry) Info(w io.Writer, specificBoxLevels, indent, indentStep string) (err error) {
+	bd := newInfoDumper(w, indent, s, -2)
+	bd.write(" * rollDistance: %d", s.RollCount)
+	bd.write(" * firstOutputSample: %d", s.FirstOutputSample)
+	level := getInfoLevel(s, specificBoxLevels)
+	if level > 0 {
+		for i, offset := range s.SampleOffset {
+			bd.write(" * sampleOffset[%d]: %d", i, offset)
```

Comment:
use i+1 instead of i here and below, since all numbering in the file format is one-based